src/CombineCodegenImpl.cxx

```diff
@@ -78,14 +78,39 @@ void RooFit::Experimental::codegenImpl(FastVerticalInterpHistPdf2& arg, CodegenC
     nominalVec[i] = cacheNominal.GetBinContent(i);
   }
 
-  // The bin index part
-  // We also have to assert that x is uniformely binned!
-  if (!dynamic_cast<RooUniformBinning const*>(&xVar.getBinning())) {
-    throw std::runtime_error("We only support uniform binning!");
+  // The bin index part - handle both uniform and non-uniform binning
+  const RooAbsBinning& binning = xVar.getBinning();
+  std::string binIdx;
+
+  if (dynamic_cast<RooUniformBinning const*>(&binning)) {
+    // UNIFORM BINNING - fast path using arithmetic
+    double xLow = xVar.getMin();
+    double xHigh = xVar.getMax();
+    binIdx = ctx.buildCall("RooFit::Detail::MathFuncs::uniformBinNumber", xLow, xHigh, xVar, numBins, 1.);
+  } else {
+    // NON-UNIFORM BINNING - use binary search
+    // Extract bin edges from binning object
+    std::vector<double> binEdges(numBins + 1);
+    for (int i = 0; i < numBins; ++i) {
+      binEdges[i] = binning.binLow(i);
+    }
+    binEdges[numBins] = binning.binHigh(numBins - 1);
+
+    // Generate code with bin edges array as static constant
+    std::string binEdgesArrayName = ctx.getTmpVarName();
+    std::stringstream binEdgesCode;
+    binEdgesCode << "static const double " << binEdgesArrayName << "[] = {";
+    for (int i = 0; i <= numBins; ++i) {
+      if (i > 0) binEdgesCode << ", ";
+      binEdgesCode << binEdges[i];
+    }
+    binEdgesCode << "};\n";
+
+    ctx.addToCodeBody(binEdgesCode.str(), true);
+
+    // Call ROOT's rawBinNumber for non-uniform bin finding
+    binIdx = ctx.buildCall("RooFit::Detail::MathFuncs::rawBinNumber", xVar, binEdgesArrayName, numBins + 1);
```

Comment:
```suggestion
    binIdx = ctx.buildCall("RooFit::Detail::MathFuncs::rawBinNumber", xVar, binEdges, numBins + 1);
```
Can you try to use the `std::vector<double> binEdges` directly to build the function call here?
The code generation context will generate the code with the copied values implicitly then, and also take care of passing the right variable name.